mfsmaster/chunks.c

```diff
@@ -3521,28 +3521,68 @@ void chunk_do_jobs(chunk *c,uint16_t scount,uint16_t fullservers,uint32_t now,ui
 						}
 					}
 */
+
 					for (i=0 ; i<dstservcnt && canbefixed ; i++) {
 						if (matching[i+labelcnt]>=0 || allowallservers) {
-							uint32_t r;
-							if (rgvc>0) {	// if there are VALID copies then make copy of one VALID chunk
-								r = 1+rndu32_ranged(rgvc);
-								srccsid = MAXCSCOUNT;
-								for (s=c->slisthead ; s && r>0 ; s=s->next) {
-									if (matocsserv_replication_read_counter(cstab[s->csid].ptr,now)<MaxReadRepl[lclass] && s->valid==VALID) {
-										r--;
-										srccsid = s->csid;
+							uint32_t min_dist = 0xFFFFFFFF;
+							uint32_t dist;
+							uint32_t ip;
+							uint32_t cuip;
+							uint16_t port;
+							uint32_t mdrgvc = 0;
+							uint32_t mdrgtdc = 0;
+							uint32_t r = 0;
+
+							srccsid = MAXCSCOUNT;
+
+							if (matocsserv_get_csdata(cstab[servers[i]].ptr,&cuip,&port,NULL,NULL)==0) {
+								for (s=c->slisthead ; s ; s=s->next) {
+									if (matocsserv_replication_read_counter(cstab[s->csid].ptr,now)<MaxReadRepl[lclass] && (s->valid==VALID || s->valid==TDVALID)) {
+										if (matocsserv_get_csdata(cstab[s->csid].ptr,&ip,&port,NULL,NULL)==0) {
+											dist=topology_distance(ip,cuip);
+											if (min_dist>=dist) {
+												if (min_dist>dist) {
+													min_dist=dist;
+													srccsid=s->csid;
+													mdrgvc = 0;
+													mdrgtdc = 0;
+												} else if (s->valid==VALID) {
+													srccsid=s->csid;
+												}
+												if (s->valid==VALID) {
+													mdrgvc++;
+												} else {
+													mdrgtdc++;
+												}
+											}
+										}
 									}
 								}
-							} else {	// if not then use TDVALID chunks.
-								r = 1+rndu32_ranged(rgtdc);
-								srccsid = MAXCSCOUNT;
+								if (mdrgvc > 1) {
+									r = 1+rndu32_ranged(mdrgvc);
+								} else if (mdrgvc == 0 && mdrgtdc > 1) {  // we have to choose TDVALID chunks
+									r = 1+rndu32_ranged(mdrgtdc);
+								}
 								for (s=c->slisthead ; s && r>0 ; s=s->next) {
-									if (matocsserv_replication_read_counter(cstab[s->csid].ptr,now)<MaxReadRepl[lclass] && s->valid==TDVALID) {
-										r--;
-										srccsid = s->csid;
+									if (matocsserv_replication_read_counter(cstab[s->csid].ptr,now)<MaxReadRepl[lclass] && (s->valid==VALID || s->valid==TDVALID)) {
+										if (matocsserv_get_csdata(cstab[s->csid].ptr,&ip,&port,NULL,NULL)==0) {
+											dist=topology_distance(ip,cuip);
+											if (min_dist==dist) {
+												if (mdrgvc > 1 && s->valid==VALID) {
+													r--;
+													srccsid=s->csid;
+												} else if (mdrgtdc > 1 && s->valid==TDVALID) {
+													r--;
+													srccsid=s->csid;
+												}
+											}
+										}
 									}
 								}
+							} else {
+								syslog(LOG_WARNING,"chunk %016"PRIX64"_%08"PRIX32" replicate failed: error get csdata of the dest chunkserver (%s)",c->chunkid,c->version,matocsserv_getstrip(cstab[servers[i]].ptr));
```

Comment:
It's normal to have temporary disconnected chunkserver, there could be flood of this. It's OK to skip them.

Reply:
If `servers[i]` is a disconnected chunkserver, then it will be removed from chunk's `slist` by `chunk_remove_diconnected_chunks`. So I think it may not cause "flood", but some warning logs in a very small possibility.
However, if the `matocsserv_get_csdata` will not get internal errors in the function itself, I also think we can remove the warning.